Interpreter.py

```diff
@@ -0,0 +1,158 @@
+            else:
+                decl.accept(self, bindings)
+    
+    def evaluate_boolean_expression(self, node, bindings):
+        left = node.left.accept(self, bindings)
+        right = node.right.accept(self, bindings)
+        
+        if node.op == "==":
+            return int(left == right)
+        elif node.op == "!=":
+            return int(left != right)
+        elif node.op == ">=":
+            return int(left >= right)
+        elif node.op == ">":
+            return int(left > right)
+        elif node.op == "<=":
+            return int(left <= right)
+        else:
+            return int(left < right)
+    
+    def evaluate_print_func(self, node, bindings):
+        list = node.listOfList[0]
+        printLine = ""
+        for expr in list:
+            printLine += str(expr.accept(self, bindings))
+            printLine += "|"
+        printLine = printLine[:-1]
+        print("Print:",printLine)
+        return printLine
+        
+    def evaluate_code_block(self, node, bindings, fn_decl):
+        lastLine = node.list[-1]
+        
+        if fn_decl == True:
+            return lastLine
+        
+        if len(node.list) > 1:
+            for line in node.list:
+                line.accept(self, bindings)
+                
+        return lastLine.accept(self, bindings)
+        
+    def evaluate_if_statement(self, node, bindings):
+        boolVal = node.boolExpr.accept(self, bindings)
+        if boolVal == 1:
+            return node.ifBlock.accept(self, bindings)
+        return node.elseBlock.accept(self, bindings)
+    
+    ###################
+    #Cusotom Functions#
+    ###################
+    
+    def evaluate_fn_decl(self, node, bindings):
+        #Handles functions declared from a closure
+        if isinstance(node.paramList, str):
+            codeBlock = bindings.getVal(node.codeBlock.name)
+            paramsToPass = node.codeBlock.paramList
+            paramsToSet = codeBlock[0]
+            
+            funcName = node.name.ident
+            funcToRun = node.codeBlock.accept(self, bindings)
+            funcToRunParamList = funcToRun.paramList
+            bindingToPass = Bindings(codeBlock[-1].binding, {})
+            
+            index = 0
+            for param in paramsToPass:
+                bindingToPass.setVal(paramsToSet[index], param.accept(self,bindings))
+                index += 1
+            
+            bindings.setFunc(funcName, funcToRunParamList, funcToRun, bindingToPass)
+            return codeBlock[1].accept(self, bindings)
+        else:
+            #Handles annonymous functions
+            if isinstance(node.name, str):
+                return node
+            #Handles normal function declaration
+            else:
+                newBindings = Bindings(bindings, {})
+                bindings.setFunc(node.name.ident, node.paramList, node.codeBlock, newBindings)
+                return node.codeBlock.accept(self, newBindings, True)
```

Comment:
Continuing the comment on the way the change to the grammar has complicated things, see how this function has extra complication?

Also, I _think_ it only handles one level of closure (although I may be wrong: it's hard to see what's going on). For example, this blows up:

~~~
let top = fn (a) {
  fn (b) {
    fn (c) {
      a + b + c
    }
  }
}

let add2 = top(2)
let add5 = add2(3)
print(add5(4))
~~~

# -2